File: python_projects/Advanced_basic/tic_tac_toe.py
def print_board(board):
    print()
    for row in board:
        print(" | ".join(row))
        print("-" * 5)

def check_winner(board):
    # Check rows, cols and diagonals
    for i in range(3):
        if board[i][0] == board[i][1] == board[i][2] != " ":
            return True
        if board[0][i] == board[1][i] == board[2][i] != " ":
            return True
    if board[0][0] == board[1][1] == board[2][2] != " ":
        return True
    if board[0][2] == board[1][1] == board[2][0] != " ":
        return True
    return False

board = [[" "]*3 for _ in range(3)]
turn = "X"
moves = 0

print("⭕❌ Welcome to Tic-Tac-Toe!")

while moves < 9:
    print_board(board)
    row = int(input(f"{turn}'s turn - Enter row (0-2): "))
    col = int(input("Enter column (0-2): "))

    if board[row][col] == " ":
        board[row][col] = turn
        moves += 1

        if check_winner(board):
            print_board(board)
            print(f"🎉 {turn} wins!")
            break

        turn = "O" if turn == "X" else "X"
    else:
        print("❌ Cell already taken!")

if moves == 9 and not check_winner(board):
    print_board(board)
    print("😐 It's a draw!")

File: java_projects/Advanced_basic/TicTacToe.java
import java.util.Scanner;

public class TicTacToe {
    public static void printBoard(char[][] board) {
        System.out.println();
        for (char[] row : board) {
            for (char cell : row) System.out.print(cell + " | ");
            System.out.println("\n---------");
        }
    }

    public static boolean checkWin(char[][] b, char mark) {
        for (int i = 0; i < 3; i++) {
            if (b[i][0] == mark && b[i][1] == mark && b[i][2] == mark) return true;
            if (b[0][i] == mark && b[1][i] == mark && b[2][i] == mark) return true;
        }
        return (b[0][0] == mark && b[1][1] == mark && b[2][2] == mark) ||
               (b[0][2] == mark && b[1][1] == mark && b[2][0] == mark);
    }

    public static void main(String[] args) {
        char[][] board = {
            {' ', ' ', ' '},
            {' ', ' ', ' '},
            {' ', ' ', ' '}
        };

        Scanner sc = new Scanner(System.in);
        char player = 'X';
        int moves = 0;

        while (moves < 9) {
            printBoard(board);
            System.out.println("Player " + player + "'s turn.");
            System.out.print("Enter row (0-2): ");
            int row = sc.nextInt();
            System.out.print("Enter col (0-2): ");
            int col = sc.nextInt();

            if (board[row][col] != ' ') {
                System.out.println("Cell already taken!");
                continue;
            }

            board[row][col] = player;
            moves++;

            if (checkWin(board, player)) {
                printBoard(board);
                System.out.println("Player " + player + " wins!");
                return;
            }

            player = (player == 'X') ? 'O' : 'X';
        }

        printBoard(board);
        System.out.println("It's a draw!");
        sc.close();
    }
}
